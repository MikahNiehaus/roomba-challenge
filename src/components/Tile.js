import React, { Component } from 'react';
import CleanTile from '../images/CleanTile.jpg';
import DirtyTile from '../images/DirtyTile.jpg';
import Roomba from '../images/Roomba.jpeg';

class Tile extends Component {
    constructor() {
        super();
        this.state = {
            roomba: false,
            dirt: false,
            id: {
                x: null,
                y: null
            }
        }
    }

    componentDidMount() {
        this.setId();
        this.checkIfDirty();
        this.checkIfRoombaIsHere();
    }

    setId() {
        const id = this.state.id;
        id["x"] = this.props.x;
        id["y"] = this.props.y;

        this.setState({id})
    }

    checkIfDirty() {
        if(this.stringify(this.props.dirt) === this.stringify(this.state.id)) {
            this.setState({dirt: true})
        }
    }

    checkIfRoombaIsHere() {
        if (this.stringify(this.state.id) === this.stringify(this.props.roomba)) {
            this.setState({dirt: false, roomba: true})
        }
    }

    stringify(object) {
        return JSON.stringify(object)
    }

    render() {

        const roomba = this.state.roomba;

        if (this.state.dirt && !roomba) {
            return (
                <div className="parentOfCleanTileWithImageInside">
                    <div>
                        <img src={CleanTile} width="100px" alt="clean looking flooring tile" className="cleanTile"/>
                        <img src={DirtyTile} width="100px" alt="dirty looking flooring tile" className="imageOnTile"/>
                    </div>
                </div>
            );
        }

        if (roomba) {
            return (
                <div>
                    <div className="parentOfCleanTileWithImageInside">
                        <img src={CleanTile} width="100px" alt="clean looking flooring tile" className="cleanTile"/>
                        <img src={Roomba} width="100px" alt="Roomba" className="imageOnTile"/>
                    </div>
                </div>
            )
        }

        else {
            return (
                <div>
                    <img src={CleanTile} width="100px" alt="clean looking flooring tile"/>
                </div>
            )

        }

    }
}

export default Tile;
