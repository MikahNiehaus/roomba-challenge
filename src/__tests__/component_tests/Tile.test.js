import React from 'react';
import {shallow} from 'enzyme';
import Tile from '../../components/Tile';

let tile;

const state = {
    roomba: false,
    dirt: false,
    id: {
        x: null,
        y: null
    }
};

describe('Tile', () => {

    beforeEach(() => {
        const props = {
            x: 1,
            y: 1,
            dirt: {
                x: 1,
                y: 2
            },
            roomba: {
                x: 0,
                y: 0
            },
        };
        tile = shallow(<Tile {...props}/>);
    });

    it('renders correctly', () => {
        expect(tile).toMatchSnapshot();
    });

    describe('Initializes state', () => {
        it('initializes state correctly', () => {
            expect(tile.state()).toEqual(state);
        });
    });

    describe('Updating the state', () => {

        describe('setID', () => {
            it('sets the ID as x=1 and y=1', () => {
                tile.instance().setId();
                expect(tile.state().id).toEqual({x: 1, y: 1})
            });
        });

        describe('checkIfDirty', () => {
            it('sets the dirt state as true with tile ID matching dirt ID', () => {
                tile.setState({id: {x: 1, y:2}});
                tile.instance().checkIfDirty();
                expect(tile.state().dirt).toEqual(true);
            });

            it('does not change the dirt state when tile ID does not match dirt ID', () => {
                tile.setState({id: {x: 2, y:2}});
                tile.instance().checkIfDirty();
                expect(tile.state().dirt).toEqual(false);
            });
        });

        describe('checkIfRoombaIsHere', () => {
            it('does not change roomba state when its ID does not match tile ID', () => {
                tile.instance().checkIfRoombaIsHere();
                expect(tile.state().roomba).toEqual(false);
            });

            it('changes roomba to true and dirt to false when its ID matches tile ID', () => {
                tile.setState({id: {x: 0, y: 0}, dirt: true});
                tile.instance().checkIfRoombaIsHere();
                expect(tile.state().roomba).toEqual(true);
                expect(tile.state().dirt).toEqual(false);
            });
        });
    });
});