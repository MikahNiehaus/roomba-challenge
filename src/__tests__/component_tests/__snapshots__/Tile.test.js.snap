// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tile renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Tile
    dirt={
        Object {
            "x": 1,
            "y": 2,
          }
    }
    roomba={
        Object {
            "x": 0,
            "y": 0,
          }
    }
    x={1}
    y={1}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <img
        alt="clean looking flooring tile"
        src="CleanTile.jpg"
        style={null}
        width="100px"
/>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "alt": "clean looking flooring tile",
        "src": "CleanTile.jpg",
        "style": null,
        "width": "100px",
      },
      "ref": null,
      "rendered": null,
      "type": "img",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <img
          alt="clean looking flooring tile"
          src="CleanTile.jpg"
          style={null}
          width="100px"
/>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "clean looking flooring tile",
          "src": "CleanTile.jpg",
          "style": null,
          "width": "100px",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
